### PR TITLE
fontbe/os2: set us{Weight,Width}Class to wght/wdth axis defaults

### LINKED ISSUE #904: should update OS/2 {weight,width}Class and post.italicAngle to match wght/wdth/slnt defaults
when fonttools varLib builds a VF, it ensures that the OS/2 weight and width classes and post.italicAngle match respectively the [wght](https://learn.microsoft.com/en-us/typography/opentype/spec/dvaraxistag_wght), [wdth](https://learn.microsoft.com/en-us/typography/opentype/spec/dvaraxistag_wdth) and [slnt](https://learn.microsoft.com/en-us/typography/opentype/spec/dvaraxistag_slnt) registered axes' default values from fvar:

https://github.com/fonttools/fonttools/blob/0c2652011e51bd78d153b41287bb845530dd0fdf/Lib/fontTools/varLib/__init__.py#L1016-L1038

https://github.com/fonttools/fonttools/blob/0c2652011e51bd78d153b41287bb845530dd0fdf/Lib/fontTools/varLib/__init__.py#L1237-L1239

Right now, fontc doesn't even set the OS/2 weight/width class (it does set the post.italicAngle).

### PR DESCRIPTION
**Pull request by anthrotype**
Fixes #904

### CODE REVIEW

File: fontbe/src/os2.rs [hunk 1]
```diff
@@ -813,6 +816,25 @@ impl Work<Context, AnyWorkId, Error> for Os2Work {
...
+            .map(|axis| axis.default.into_inner().0)
+            .unwrap_or(400.0);
+        let us_weight_class: u16 = default_wght.clamp(1.0, 1000.0).ot_round();
+
+        let default_wdth = static_metadata
+            .axes
+            .iter()
+            .find(|axis| axis.tag == Tag::new(b"wdth"))
+            .map(|axis| axis.default.into_inner().0)
+            .unwrap_or(100.0);
```
Comment (rsheeter) [inline fontbe/src/os2.rs:835]: These lines repeat almost identically here and above. I suggest extracting a function or even adding it to static metadata, e.g. `static_metadata.axis(Tag) -> Option<Axis>`

File: fontdrasil/src/types.rs [hunk 2]
```diff
@@ -156,4 +157,18 @@ impl WidthClass {
             WidthClass::UltraExpanded => 200.0,
         }
     }
+
+    pub fn from_percent(percent: f32) -> Self {
```
Comment (rsheeter) [inline fontdrasil/src/types.rs:161]: Should document what happens if pct doesn't exactly match a class. Changing the name might suffice, e.g. `nearest_class(pct: f32)` I'm immediately clear what I get.

File: fontdrasil/src/types.rs [hunk 3]
```diff
@@ -156,4 +157,18 @@ impl WidthClass {
...
+        let mappings = WidthClass::all_values()
+            .iter()
+            .map(|wc| (wc.to_percent().into(), (*wc as u16).into()))
+            .collect();
+        let v: u16 = PiecewiseLinearMap::new(mappings)
```
Comment (rsheeter) [inline fontdrasil/src/types.rs:166]: I don't think I'd have expected this to allocate. 

This is testable, I suggest adding tests including out of bounds values. At a glance I'm slightly skeptical this works if the value if out of bounds.

Comment (rsheeter) [reply fontdrasil/src/types.rs:166]: I tried it, `WidthClass::from_percent(42.0)` fails (bonus points: not the way I expected  ) with `attempt to subtract with overflow`

Comment (rsheeter) [reply fontdrasil/src/types.rs:166]: Over-large values also fail.

Comment (rsheeter) [reply fontdrasil/src/types.rs:166]: Exploratory fix (PR to this branch) in https://github.com/googlefonts/fontc/pull/923

**Review by rsheeter — APPROVED** [review]
LGTM with a couple of suggestions

File: fontdrasil/src/types.rs (hunk 3 above)
Comment (anthrotype) [reply fontdrasil/src/types.rs:166]: but why doesn't it work? I thought PiecewiseLinearMap worked the same way as fonttools' equivalent. I used the same function that varLib uses for the same task and naively assumed they'd work the same. Should they not work the same?

**Comment by anthrotype** [comment]
as for post.italicAngle, fontbe already sets that one, I don't think we need to do anything about it. The font developer should already ensure that it matches the slnt default.

> In a variable font that implements 'slnt' variations, the value in the italicAngle field of the 'post' table must match the default 'slnt' value specified in the ['fvar'](https://learn.microsoft.com/en-gb/typography/opentype/spec/fvar) table.

https://learn.microsoft.com/en-gb/typography/opentype/spec/dvaraxistag_slnt

File: fontdrasil/src/types.rs (hunk 3 above)
Comment (rsheeter) [reply fontdrasil/src/types.rs]: Fair point, the non-handling of our of bounds values should be viewed as a bug then. I will file an issue. I still think we prefer not to allocate here.

Comment (rsheeter) [reply fontdrasil/src/types.rs]: https://github.com/googlefonts/fontc/issues/924

**Pull request merged**